lib/src/main/java/org/exoplatform/social/addons/updater/MigrationIdentityStorage.java

```diff
@@ -0,0 +1,189 @@
+import org.exoplatform.social.core.storage.IdentityStorageException;
+import org.exoplatform.social.core.storage.api.IdentityStorage;
+import org.exoplatform.social.core.storage.impl.IdentityStorageImpl;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class is introduced because there are many other service which start and run in parallel with migration service.
+ * These services will load/create identity via api @{IdentityManger#getOrCreateIdentity(String providerId, String remoteId, boolean isProfileLoaded)}
+ *
+ * This class will be used in replace of RDBMSIdentityStorageImpl during the migration time.
+ * It will take care to migrate all identity which is used before the migration service migrate that one
+ *
+ * @author <a href="mailto:tuyennt@exoplatform.com">Tuyen Nguyen The</a>.
+ */
+public class MigrationIdentityStorage implements IdentityStorage {
+
+  private final IdentityStorageImpl jcrStorage;
+  private final RDBMSIdentityStorageImpl jpaStorage;
+  private final IdentityMigrationService migrationService;
+
+  public MigrationIdentityStorage(IdentityStorageImpl jcrStorage, RDBMSIdentityStorageImpl jpaStorage, IdentityMigrationService migrationService) {
+    this.jcrStorage = jcrStorage;
+    this.jpaStorage = jpaStorage;
+    this.migrationService = migrationService;
+  }
+
+  @Override
+  public void saveIdentity(Identity identity) throws IdentityStorageException {
+    jpaStorage.saveIdentity(identity);
+  }
+
+  @Override
+  public Identity updateIdentity(Identity identity) throws IdentityStorageException {
+    return jpaStorage.updateIdentity(identity);
+  }
+
+  @Override
+  public void updateIdentityMembership(String remoteId) throws IdentityStorageException {
+    jpaStorage.updateIdentityMembership(remoteId);
+  }
+
+  @Override
+  public Identity findIdentityById(String nodeId) throws IdentityStorageException {
+    Identity identity = jpaStorage.findIdentityById(nodeId);
+    if (identity == null) {
+      if (jcrStorage.findIdentityById(nodeId) != null) {
+        identity = migrationService.migrateIdentity(nodeId);
+      }
+    }
+    return identity;
+  }
+
+  @Override
+  public void deleteIdentity(Identity identity) throws IdentityStorageException {
+    jpaStorage.deleteIdentity(identity);
+  }
+
+  @Override
+  public void hardDeleteIdentity(Identity identity) throws IdentityStorageException {
+    jpaStorage.hardDeleteIdentity(identity);
+  }
+
+  @Override
+  public Profile loadProfile(Profile profile) throws IdentityStorageException {
+    return jpaStorage.loadProfile(profile);
+  }
+
+  @Override
+  public Identity findIdentity(String providerId, String remoteId) throws IdentityStorageException {
+    Identity identity = jpaStorage.findIdentity(providerId, remoteId);
+    if (identity == null) {
+      identity = jcrStorage.findIdentity(providerId, remoteId);
+      if (identity != null) {
+        identity = migrationService.migrateIdentity(identity.getId());
```

Comment:
In this case, one of these migrations will be failed as the transaction commit failure (as the exception reported in jira issue) but this identity still migrated successfully (because the remain migration will commit successfully)